### PR TITLE
backend_cairo: Clip drawn paths to context.clip_extents()

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
+ Fixes https://github.com/matplotlib/matplotlib/issues/4054

### CODE REVIEW

**Comment by MathieuDuponchelle** [comment]
Using clip_to_bbox was wrong, as behaviour is undefined with unclosed paths, indentation seems correct to me, and "translate" the clip extents to the value range.

**Comment by MathieuDuponchelle** [comment]
Not sure why the build failed, seems like pep8 isn't happy with other files afaict

File: lib/matplotlib/backends/backend_cairo.py [hunk 1]
```diff
@@ -132,8 +132,18 @@ def _fill_and_stroke (self, ctx, fill_c, alpha, alpha_overrides):
...
+    def convert_path(self, ctx, path, transform):
+        # First clip the path to the actual rendering extents
+        points = path.get_extents().get_points()
+        width = points[1][0] - points[0][0]
+        height = points[1][1] - points[0][1]
+        wfactor = width / self.width
+        hfactor = height / self.height
+
+        extents = ctx.clip_extents()
+        clip_rect = (extents[0] * wfactor, extents[1] * hfactor, extents[2] * wfactor, extents[3] * hfactor)
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: This line is too long.

File: lib/matplotlib/backends/backend_cairo.py [hunk 2]
```diff
@@ -132,8 +132,18 @@ def _fill_and_stroke (self, ctx, fill_c, alpha, alpha_overrides):
...
 
-    @staticmethod
-    def convert_path(ctx, path, transform):
+    def convert_path(self, ctx, path, transform):
+        # First clip the path to the actual rendering extents
+        points = path.get_extents().get_points()
+        width = points[1][0] - points[0][0]
+        height = points[1][1] - points[0][1]
+        wfactor = width / self.width
+        hfactor = height / self.height
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: I don't understand what this is doing.

Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: This is some logic to translate the cairo clip_extents to the actual path extents.

Let's say that
1) we draw on a canvas that is 1000 pixels wide (self.width)
2) The paths we're drawing have values ranging from 0 to 100000 "data units"
3) the region that we have to draw on (ctx.clip_extents()) is from 0 *pixels* to 100 *pixels*
4) We thus want to clip the path to 0 to 10000 "data units", ie region_width times "data width" divided by self.width (100 * 100000 / 1000)

I hope I've been clear enough, there may be other ways to achieve this but I am quite sure this one is correct now (compared to the previous proposal).

File: lib/matplotlib/backends/backend_cairo.py (hunk 1 above)
Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: OK, will change once you've acked my other comment, I think amending commits and pushing kills line notes

File: lib/matplotlib/backends/backend_cairo.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: This is what the transform stack is for: http://matplotlib.org/users/transforms_tutorial.html?highlight=transform%20tutorial

Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: Not sure how to rescale these extents from display units to data units, any hint ?

**Comment by MathieuDuponchelle** [comment]
Updated the patch to use the transform stack, no more lines exceeding 80 columns

**Comment by MathieuDuponchelle** [comment]
Can this be merged as part of the 1.4.X series ?

**Comment by mdboom** [comment]
Doesn't this also have the issue with filled paths?  The path clipping algorithm in matplotlib does not support filled paths, so it should be turned on only for unfilled paths -- but it looks like this is clipping always.

**Comment by tacaswell** [comment]
I don't think this will go into the 1.4.x series as 1.4.3 already has a release candidate out and there is not a plan for a 1.4.4.

**Comment by MathieuDuponchelle** [comment]
@mdboom , by filled paths do you mean "unclosed paths" ?

**Comment by mdboom** [comment]
I mean paths with a fill (as opposed to just stroked around the edge).

**Comment by mdboom** [comment]
In this case, `if rgbFace` is not None or gc.get_hatch() is not None` then you can do the clipping.

**Comment by MathieuDuponchelle** [comment]
Thanks, I'll update the patch

**Comment by MathieuDuponchelle** [comment]
@mdboom, in my example, I both have rgbFace and gc.get_hatch() to None and my patch works, I don't think this path is filled anyway, are you sure about your last comment?

**Comment by MathieuDuponchelle** [comment]
There indeed is an issue when replacing the call to plot with a call to fill, but in both cases gc.get_hatch() and rbgFace are None

**Comment by MathieuDuponchelle** [comment]
I was mistaken actually, I think you might have been too, the condition is "if rgbFace is not None or gc.get_hatch() is not None then I *can't* do the clipping", cause in the case of fill I indeed have a rgbFace.

I'll update my patch

**Comment by MathieuDuponchelle** [comment]
Hey, can we go ahead with that patch? I'm not sure why the checks fail here to be honest?

**Comment by MathieuDuponchelle** [comment]
OK, I needed to rebase my branch on master to make the checks pass, that's done. I have made all the changes that were pointed out, would be great to merge this now :)

**Comment by mdboom** [comment]
Sorry to not follow up on this patch earlier.  This looks fine, and the logic about when to do the clipping is correct.  However, now `path.cleanup` (which is a rather computationally intensive task) is being called twice: once in `draw_path`, and once in `Path.iter_segments` within `convert_path`.  I think it would be better to pass `clip` down into `iter_segments` to actually do the clipping (and passing `None` will disable clipping).  So something like:

```
if rgbFace is None and gc.get_hatch() is None:
    clip = ctx.clip_extents()
else:
    clip = None
```

and then modify `convert_path` so it takes a clip argument and passes it along to `iter_segments`.  Also no need to convert the extents to data space first in this scenario.

**Comment by MathieuDuponchelle** [comment]
Unfortunately, iter_segments with clip=clip didn't clip as expected, that's why I did it in a separate step.

**Comment by MathieuDuponchelle** [comment]
You can see this by using the script pasted in #4075 and modifying my patch by the way

**Comment by MathieuDuponchelle** [comment]
Ahem, actually it does work, yay :D I'm not sure why it didn't work first time I tried, but it makes the patch much cleaner, thanks

File: lib/matplotlib/backends/backend_cairo.py [hunk 3]
```diff
@@ -133,8 +133,8 @@ def _fill_and_stroke (self, ctx, fill_c, alpha, alpha_overrides):
         ctx.stroke()
 
     @staticmethod
-    def convert_path(ctx, path, transform):
-        for points, code in path.iter_segments(transform):
+    def convert_path(ctx, path, transform, clip):
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_cairo.py]: `clip` should be optional or this is an API break.

Comment (MathieuDuponchelle) [inline lib/matplotlib/backends/backend_cairo.py]: Fixed

**Comment by MathieuDuponchelle** [comment]
ping :)

**Comment by tacaswell** [comment]
Thanks for the patch and bearing with us through the review process!

**Comment by MathieuDuponchelle** [comment]
No problem, glad it made it in, sorry about the rookie mistakes but I'm really not familiar with the codebase :)

**Pull request merged**